src/gpl/src/initialPlace.cpp

```diff
@@ -106,23 +142,157 @@ void InitialPlace::doBicgstabPlace() {
+      CUDA_ERROR(cudaMemcpy(d_fixedInstForceVecY_, fixedInstForceVecY_.data(), sizeof(float)*m, cudaMemcpyHostToDevice));
+      
+      // Set handler
+      cusolverSpHandle_t handleCusolverX = NULL;
+      cusparseHandle_t handleCusparseX = NULL;
+      cudaStream_t streamX = NULL;
+
+      cusolverSpHandle_t handleCusolverY = NULL;
+      cusparseHandle_t handleCusparseY = NULL;
+      cudaStream_t streamY = NULL;
+
+      // Initialize handler
+      CUSOLVER_ERROR(cusolverSpCreate(&handleCusolverX));
+      CUSPARSE_ERROR(cusparseCreate(&handleCusparseX));
+      CUDA_ERROR(cudaStreamCreate(&streamX));
+      CUSOLVER_ERROR(cusolverSpSetStream(handleCusolverX, streamX));
+      CUSPARSE_ERROR(cusparseSetStream(handleCusparseX, streamX));
+
+      CUSOLVER_ERROR(cusolverSpCreate(&handleCusolverY));
+      CUSPARSE_ERROR(cusparseCreate(&handleCusparseY));
+      CUDA_ERROR(cudaStreamCreate(&streamY));
+      CUSOLVER_ERROR(cusolverSpSetStream(handleCusolverY, streamY));
+      CUSPARSE_ERROR(cusparseSetStream(handleCusparseY, streamY));
+
+      // Create and define cusparse descriptor
+      cusparseMatDescr_t descrX = NULL;
+      CUSPARSE_ERROR(cusparseCreateMatDescr(&descrX));
+      CUSPARSE_ERROR(cusparseSetMatType(descrX, CUSPARSE_MATRIX_TYPE_GENERAL));
+      CUSPARSE_ERROR(cusparseSetMatIndexBase(descrX, CUSPARSE_INDEX_BASE_ZERO));
+            
+      cusparseMatDescr_t descrY = NULL;
+      CUSPARSE_ERROR(cusparseCreateMatDescr(&descrY));
+      CUSPARSE_ERROR(cusparseSetMatType(descrY, CUSPARSE_MATRIX_TYPE_GENERAL));
+      CUSPARSE_ERROR(cusparseSetMatIndexBase(descrY, CUSPARSE_INDEX_BASE_ZERO));
+
+      // transform from coordinates (COO) values to compressed row pointers (CSR) values
+      // https://docs.nvidia.com/cuda/cusparse/index.html
+      int *d_csrRowIndX = NULL, *d_csrRowIndY = NULL;;
+      CUDA_ERROR(cudaMalloc((void**)&d_csrRowIndX, (m+1) * sizeof(int)));
+      CUDA_ERROR(cudaMalloc((void**)&d_csrRowIndY, (m+1) * sizeof(int)));
+      CUSPARSE_ERROR(cusparseXcoo2csr(handleCusparseX, d_cooRowIndexX, nnz_x, m, d_csrRowIndX, CUSPARSE_INDEX_BASE_ZERO));
+      CUSPARSE_ERROR(cusparseXcoo2csr(handleCusparseY, d_cooRowIndexY, nnz_y, m, d_csrRowIndY, CUSPARSE_INDEX_BASE_ZERO));
+      CUSOLVER_ERROR(cusolverSpScsrlsvqr(handleCusolverX, m, nnz_x, descrX, d_cooValX, d_csrRowIndX, d_cooColIndexX, d_fixedInstForceVecX_, tol, reorder, d_instLocVecX, &singularity));
+      CUSOLVER_ERROR(cusolverSpScsrlsvqr(handleCusolverY, m, nnz_y, descrY, d_cooValY, d_csrRowIndY, d_cooColIndexY, d_fixedInstForceVecY_, tol, reorder, d_instLocVecY, &singularity));
+      
+      // Sync and Copy data to host
+      CUDA_ERROR(cudaMemcpyAsync(instLocVecX_.data(), d_instLocVecX, sizeof(float)*m, cudaMemcpyDeviceToHost, streamX));
+      CUDA_ERROR(cudaMemcpyAsync(instLocVecY_.data(), d_instLocVecY, sizeof(float)*m, cudaMemcpyDeviceToHost, streamY));
+
+      // Destroy what is not needed in both of device and host
```

Comment:
Just want to check -- this mentions destroying what is not needed, all of the on-device constructs will be allocated and freed every iteration, correct?

Reply:
Yes.